### DISCUSSION TITLE
How to extract data from an event?

### DISCUSSION THREAD

**Discussion by SamuelLouisCampbell** [cesanta/mongoose#2130] [Ask questions, get answers] [answered]
Apologies, my C++ is rusty and my C is non existent but if some generous person could hold my hand....
I'm trying to create a little C++ wrapper for mongoose so I can integrate it into one of my old projects, I am stuck however on how one might extract data from the static user function so I can pass it on up out of my wrapper. 

My user function is defined like so in my .h file;
`	static void fn(mg_connection* c, int ev, void* ev_data, void* fn_data);
`
and then in the cpp file;
```
	void MgsWrapper::fn(mg_connection* c, int ev, void* ev_data, void* fn_data)
	{
		struct mg_http_serve_opts opts;
		opts.extra_headers = NULL;
		opts.fs = NULL;
		opts.mime_types = NULL;
		opts.page404 = NULL;
		opts.root_dir = ".";
		opts.ssi_pattern = NULL;
	
		if (ev == MG_EV_HTTP_MSG)
		{
			mg_http_serve_dir(c, static_cast<mg_http_message*>(ev_data), &opts);
			struct mg_http_message* hm = static_cast<mg_http_message*>(ev_data);
			char buffer[256] = "";
			int len = mg_http_get_var(&hm->body, "data", buffer, sizeof(buffer));
			if (len > 0)
			{
				std::string m;                                           //How can I make the data in 'm' available outside of this function?      
				m = convertToString(buffer, len);             //If I try to define a string outside of this scope and fill it, I get errors as 
				std::cout << m.message;                        // this is a static function,
			}
		}
	}
```
I hope that's clear, many thanks!

**Comment by scaprile** [comment, marked as answer]
Your problem is not related to Mongoose nor its events, but on general programming.
The event handler doesn't _have_ to be static, you can have a global if you want, and you can do whatever you want to do with your data; put it in a shared variable, push it to a queue, save it to a file, write to a Un*x pipe, to an RTOS mailbox, whatever. Just keep the manager and the handlers on the same thread.
